tests/unit/linalg/test_constrainedsolver.cpp

```diff
@@ -0,0 +1,347 @@
+   solver.GetDualSolution(lambda);
+   serr(0) = truex - sol(0);
+   serr(1) = truey - sol(1);
+   lerr(0) = truelambda - lambda(0);
+}
+
+// this test case is intended to run on one processor, but it is
+// marked [Parallel] because it uses hypre
+TEST_CASE("ConstrainedSolver", "[Parallel], [ConstrainedSolver]")
+{
+   int comm_size;
+   MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+
+   if (comm_size == 1)
+   {
+      Vector serr(2);
+      Vector lerr(1);
+
+      SimpleSaddle problem(4.0, -2.0);
+
+      problem.Schur(serr, lerr);
+      REQUIRE(serr(0) == MFEM_Approx(0.0));
+      REQUIRE(serr(1) == MFEM_Approx(0.0));
+      REQUIRE(lerr(0) == MFEM_Approx(0.0));
+
+      problem.Elimination(serr, lerr);
+      REQUIRE(serr(0) == MFEM_Approx(0.0));
+      REQUIRE(serr(1) == MFEM_Approx(0.0));
+      REQUIRE(lerr(0) == MFEM_Approx(0.0));
+
+      for (auto pen : {1.e+3, 1.e+4, 1.e+6})
+      {
+         problem.Penalty(pen, serr, lerr);
+         REQUIRE(std::abs(serr(0)) < 1./pen);
+         REQUIRE(std::abs(serr(1)) < 1./pen);
+         REQUIRE(std::abs(lerr(0)) < 1./pen);
+      }
+
+      Vector dualrhs(1);
+      dualrhs(0) = 1.0;
+      problem.SetDualRHS(dualrhs);
+
+      problem.Schur(serr, lerr);
+      REQUIRE(serr(0) == MFEM_Approx(0.0));
+      REQUIRE(serr(1) == MFEM_Approx(0.0));
+      REQUIRE(lerr(0) == MFEM_Approx(0.0));
+
+      problem.Elimination(serr, lerr);
+      REQUIRE(serr(0) == MFEM_Approx(0.0));
+      REQUIRE(serr(1) == MFEM_Approx(0.0));
+      REQUIRE(lerr(0) == MFEM_Approx(0.0));
+
+      for (auto pen : {1.e+3, 1.e+4, 1.e+6})
+      {
+         problem.Penalty(pen, serr, lerr);
+         REQUIRE(std::abs(serr(0)) < 1./pen);
+         REQUIRE(std::abs(serr(1)) < 1./pen);
+         REQUIRE(std::abs(lerr(0)) < 1./pen);
+      }
+   }
+}
+
+
+class ParallelTestProblem
+{
+public:
+   ParallelTestProblem();
+   ~ParallelTestProblem();
+
+   void Schur(Vector& serr, Vector& lerr);
+   void Penalty(double pen, Vector& serr, Vector& lerr);
+
+private:
+   SparseMatrix Alocal;
+   Vector rhs, sol, truesol, lambda, truelambda;
+   HypreParMatrix * amat;
+   HypreParMatrix * bmat;
+};
+
+
+ParallelTestProblem::ParallelTestProblem()
+   :
+   Alocal(2), rhs(2), sol(2), truesol(2), lambda(1), truelambda(1)
+{
+   int rank;
+   MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+   Alocal.Add(0, 0, 1.0);
+   Alocal.Add(1, 1, 1.0);
+   Alocal.Finalize();
+
+   int row_starts_a[2] = {2 * rank, 2 * (rank + 1)};
+   amat = new HypreParMatrix(MPI_COMM_WORLD, 8, row_starts_a, &Alocal);
+   amat->CopyRowStarts();
+   
+   SparseMatrix Blocal(1, 8);
+   if (rank == 3)
+   {
+      Blocal.Add(0, 0, 1.0);
+      Blocal.Add(0, 7, 1.0);
+   }
+   else
+   {
+      Blocal.Add(0, 2*rank + 1, 1.0);
+      Blocal.Add(0, 2*rank + 2, 1.0);
+   }
+   Blocal.Finalize();
+   int row_starts_c[2] = { rank, rank + 1 };
+   int col_starts[2] = { 2*rank, 2 * (rank + 1) };
+   bmat = new HypreParMatrix(MPI_COMM_WORLD, 1, 4, 8, Blocal.GetI(),
```

Comment:
I assume there's no way around a user explicitly constructing a HypreParMatrix in the parallel case. I can't think of a good way out of this, but I'm open to any ideas.

Reply:
The `SchurConstrainedSolver` takes an abstract `Operator` here, so if the user has something else in mind that's possible. But of course the more abstract the interface, the less we can do for preconditioning. It also may be possible to add a layer so the user provides (for example) a contact surface and we build the constraint `HypreParMatrix`.